packages/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/FirestoreMessageCodec.java

```diff
@@ -0,0 +1,452 @@
+        stream.write(0);
+      }
+    }
+  }
+
+  static void writeValue(ByteArrayOutputStream stream, Object value) {
+    if (value == null) {
+      stream.write(NULL);
+    } else if (value == Boolean.TRUE) {
+      stream.write(TRUE);
+    } else if (value == Boolean.FALSE) {
+      stream.write(FALSE);
+    } else if (value instanceof Number) {
+      if (value instanceof Integer || value instanceof Short || value instanceof Byte) {
+        stream.write(INT);
+        writeInt(stream, ((Number) value).intValue());
+      } else if (value instanceof Long) {
+        stream.write(LONG);
+        writeLong(stream, (long) value);
+      } else if (value instanceof Float || value instanceof Double) {
+        stream.write(DOUBLE);
+        writeAlignment(stream, 8);
+        writeDouble(stream, ((Number) value).doubleValue());
+      } else if (value instanceof BigInteger) {
+        stream.write(BIGINT);
+        writeBytes(stream, ((BigInteger) value).toString(16).getBytes(UTF8));
+      } else {
+        throw new IllegalArgumentException("Unsupported Number type: " + value.getClass());
+      }
+    } else if (value instanceof String) {
+      stream.write(STRING);
+      writeBytes(stream, ((String) value).getBytes(UTF8));
+    } else if (value instanceof byte[]) {
+      stream.write(BYTE_ARRAY);
+      writeBytes(stream, (byte[]) value);
+    } else if (value instanceof int[]) {
+      stream.write(INT_ARRAY);
+      final int[] array = (int[]) value;
+      writeSize(stream, array.length);
+      writeAlignment(stream, 4);
+      for (final int n : array) {
+        writeInt(stream, n);
+      }
+    } else if (value instanceof long[]) {
+      stream.write(LONG_ARRAY);
+      final long[] array = (long[]) value;
+      writeSize(stream, array.length);
+      writeAlignment(stream, 8);
+      for (final long n : array) {
+        writeLong(stream, n);
+      }
+    } else if (value instanceof double[]) {
+      stream.write(DOUBLE_ARRAY);
+      final double[] array = (double[]) value;
+      writeSize(stream, array.length);
+      writeAlignment(stream, 8);
+      for (final double d : array) {
+        writeDouble(stream, d);
+      }
+    } else if (value instanceof List) {
+      stream.write(LIST);
+      final List<?> list = (List) value;
+      writeSize(stream, list.size());
+      for (final Object o : list) {
+        writeValue(stream, o);
+      }
+    } else if (value instanceof Map) {
+      stream.write(MAP);
+      final Map<?, ?> map = (Map) value;
+      writeSize(stream, map.size());
+      for (final Entry entry : map.entrySet()) {
+        writeValue(stream, entry.getKey());
+        writeValue(stream, entry.getValue());
+      }
+    } else if (value instanceof Date) {
+      stream.write(DATE_TIME);
+      final Date date = (Date) value;
+      final long milliseconds = date.getTime();
+      writeLong(stream, milliseconds * 1000);
+    } else if (value instanceof GeoPoint) {
```

Comment:
There's no FieldValue encoding here as I'm not sure if there's a way to check whether it's a `serverTimestamp()` or `delete()`. This won't be a big blocker, as it should only affect cached data. FieldValues should be converted by the server, right?